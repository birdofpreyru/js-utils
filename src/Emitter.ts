export type Listener<T extends unknown[] = unknown[]> = (...args: T) => void;

/**
 * Simple listeneable data Emitter.
 */
export class Emitter<T extends unknown[] = unknown[]> {
  private pListeners: Array<Listener<T>> = [];

  /**
   * Returns "true" if any listener is connected; "false" otherwise.
   * @return {boolean}
   */
  get hasListeners(): boolean {
    return !!this.pListeners.length;
  }

  get listeners(): ReadonlyArray<Listener<T>> {
    return this.pListeners;
  }

  /**
   * Adds `listener` if it is not already connected.
   * @param {function} listener
   * @return {function} Unsubscribe function.
   */
  addListener(listener: Listener<T>): () => void {
    if (!this.pListeners.includes(listener)) {
      this.pListeners.push(listener);
    }
    return () => {
      this.removeListener(listener);
    };
  }

  /**
   * Calls every connected listener with the given arguments.
   * @param args
   */
  emit(...args: T): void {
    const listeners = this.pListeners.slice();
    for (const listener of listeners) listener(...args);
  }

  /**
   * Removes all connected listeners.
   */
  removeAllListeners(): void {
    this.pListeners = [];
  }

  /**
   * Removes specified `listener`, if connected.
   * @param listener
   */
  removeListener(listener: Listener<T>): void {
    const idx = this.pListeners.indexOf(listener);
    if (idx >= 0) this.pListeners.splice(idx, 1);
  }
}
