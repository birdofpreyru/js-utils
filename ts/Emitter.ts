type Listener = (...args: any[]) => void;

/**
 * Simple listeneable data Emitter.
 */
export default class Emitter {
  private p_listeners: Listener[] = [];

  /**
   * Returns "true" if any listener is connected; "false" otherwise.
   * @return {boolean}
   */
  get hasListeners(): boolean {
    return !!this.p_listeners.length;
  }

  get listeners(): ReadonlyArray<Listener> { return this.p_listeners; }

  /**
   * Adds `listener` if it is not already connected.
   * @param {function} listener
   * @return {function} Unsubscribe function.
   */
  addListener(listener: Listener): () => void {
    if (!this.p_listeners.includes(listener)) {
      this.p_listeners.push(listener);
    }
    return () => this.removeListener(listener);
  }

  /**
   * Calls every connected listener with the given arguments.
   * @param  {...any} args
   */
  emit(...args: any[]) {
    const { p_listeners: listeners } = this;
    for (let i = 0; i < listeners.length; ++i) {
      listeners[i](...args);
    }
  }

  /**
   * Removes specified `listener`, if connected.
   * @param {function} listener
   */
  removeListener(listener: Listener) {
    const idx = this.p_listeners.indexOf(listener);
    if (idx >= 0) this.p_listeners.splice(idx, 1);
  }
}
