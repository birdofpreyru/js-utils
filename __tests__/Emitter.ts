import Emitter from '../ts/Emitter';

describe('constructor()', () => {
  it('creates a new emitter with empty listeners array', () => {
    const e = new Emitter();
    expect(e.listeners).toEqual([]);
  });
});

describe('.hasListeners', () => {
  it('returns "false" when no listeners connected', () => {
    const e = new Emitter();
    expect(e.hasListeners).toBe(false);
  });

  it('returns "true" when a listener is connected', () => {
    const e = new Emitter();
    e.addListener(jest.fn());
    expect(e.hasListeners).toBe(true);
  });
});

describe('.addListener()', () => {
  it('adds given listeners at most once', () => {
    const e = new Emitter();
    const fnA = jest.fn().mockName('A');
    const fnB = jest.fn().mockName('B');
    e.addListener(fnA);
    expect(e.listeners).toMatchSnapshot();
    e.addListener(fnA);
    expect(e.listeners).toMatchSnapshot();
    e.addListener(fnB);
    expect(e.listeners).toMatchSnapshot();
    e.addListener(fnA);
    expect(e.listeners).toMatchSnapshot();
  });

  it('removes functional unsubscribe function', () => {
    const e = new Emitter();
    const fnA = jest.fn().mockName('A');
    const fnB = jest.fn().mockName('B');
    const fnC = jest.fn().mockName('C');
    e.addListener(fnA);
    const unB = e.addListener(fnB);
    const unC = e.addListener(fnC);
    expect(e.listeners).toMatchSnapshot();
    unB();
    expect(e.listeners).toMatchSnapshot();
    unC();
    expect(e.listeners).toMatchSnapshot();
    unC();
    expect(e.listeners).toMatchSnapshot();
  });
});

describe('.emit()', () => {
  it('calls connected listeners with given arguments', () => {
    const e = new Emitter();
    const fnA = jest.fn().mockName('A');
    const fnB = jest.fn().mockName('B');
    e.addListener(fnA);
    e.addListener(fnB);
    e.emit('ARG1', 'arg2');
    expect(fnA.mock.calls).toMatchSnapshot();
    expect(fnB.mock.calls).toMatchSnapshot();
    e.removeListener(fnA);
    e.emit('second call');
    expect(fnA.mock.calls).toMatchSnapshot();
    expect(fnB.mock.calls).toMatchSnapshot();
  });
});

describe('.removeListener()', () => {
  it('removes given listener if it is connected', () => {
    const e = new Emitter();
    const fnA = jest.fn().mockName('A');
    const fnB = jest.fn().mockName('B');
    const fnC = jest.fn().mockName('C');
    const fnD = jest.fn().mockName('D');
    e.addListener(fnA);
    e.addListener(fnB);
    e.addListener(fnC);
    e.addListener(fnD);
    expect(e.listeners).toMatchSnapshot();
    e.removeListener(fnA);
    expect(e.listeners).toMatchSnapshot();
    e.removeListener(fnC);
    expect(e.listeners).toMatchSnapshot();
    e.removeListener(fnA);
    e.removeListener(fnC);
    expect(e.listeners).toMatchSnapshot();
  });
});
