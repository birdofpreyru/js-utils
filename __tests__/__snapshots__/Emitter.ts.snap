// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.addListener() adds given listeners at most once 1`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
  ],
}
`;

exports[`.addListener() adds given listeners at most once 2`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
  ],
}
`;

exports[`.addListener() adds given listeners at most once 3`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
    [MockFunction B],
  ],
}
`;

exports[`.addListener() adds given listeners at most once 4`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
    [MockFunction B],
  ],
}
`;

exports[`.addListener() removes functional unsubscribe function 1`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
    [MockFunction B],
    [MockFunction C],
  ],
}
`;

exports[`.addListener() removes functional unsubscribe function 2`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
    [MockFunction C],
  ],
}
`;

exports[`.addListener() removes functional unsubscribe function 3`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
  ],
}
`;

exports[`.addListener() removes functional unsubscribe function 4`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
  ],
}
`;

exports[`.emit() calls connected listeners with given arguments 1`] = `
[
  [
    "ARG1",
    "arg2",
  ],
]
`;

exports[`.emit() calls connected listeners with given arguments 2`] = `
[
  [
    "ARG1",
    "arg2",
  ],
]
`;

exports[`.emit() calls connected listeners with given arguments 3`] = `
[
  [
    "ARG1",
    "arg2",
  ],
]
`;

exports[`.emit() calls connected listeners with given arguments 4`] = `
[
  [
    "ARG1",
    "arg2",
  ],
  [
    "second call",
  ],
]
`;

exports[`.removeListener() removes given listener if it is connected 1`] = `
Emitter {
  "p_listeners": [
    [MockFunction A],
    [MockFunction B],
    [MockFunction C],
    [MockFunction D],
  ],
}
`;

exports[`.removeListener() removes given listener if it is connected 2`] = `
Emitter {
  "p_listeners": [
    [MockFunction B],
    [MockFunction C],
    [MockFunction D],
  ],
}
`;

exports[`.removeListener() removes given listener if it is connected 3`] = `
Emitter {
  "p_listeners": [
    [MockFunction B],
    [MockFunction D],
  ],
}
`;

exports[`.removeListener() removes given listener if it is connected 4`] = `
Emitter {
  "p_listeners": [
    [MockFunction B],
    [MockFunction D],
  ],
}
`;

exports[`constructor() creates a new emitter with empty listeners array 1`] = `
Emitter {
  "p_listeners": [],
}
`;
